File: javascript_games/ninetyninebottles2.js
/*
niNety-nniinE BoOttels of Mlik On teh waLl
By Al Sweigart al@inventwithpython.com
Print the full lyrics to one of the longest songs ever! The song
gets sillier and with each verse. Press Ctrl-C to stop.
*/

'use strict';

function sleep(seconds) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

// Set up the constants:
// (!) Try changing both of these to 0 to print all the lyrics at once.
const SPEED = 0.01;  // The pause in between printing letters.
const LINE_PAUSE = 1.5;  // The pause at the end of each line.


async function slowPrint(text, pauseAmount=0.1) {
    for (let i = 0; i < text.length; i++) {
        process.stdout.write(text[i]);
        await sleep(pauseAmount)  // Pause in between each character.
    }
    console.log()  // Print a newline.
}


function replaceAt(str, index, replacement) {
    return str.substr(0, index) + replacement + str.substr(index + replacement.length);
}


async function main() {
    console.log('niNety-nniinE BoOttels, by Al Sweigart al@inventwithpython.com');
    console.log();
    console.log('(Press Ctrl-C to quit.)');

    await sleep(2);

    let bottles = 99;  // This is the starting number of bottles.

    // This list holds the string used for the lyrics:
    let lines = [' bottles of milk on the wall,',
             ' bottles of milk,',
             'Take one down, pass it around,',
             ' bottles of milk on the wall!'];

    while (bottles > 0) {  // Keep looping and display the lyrics.
        await slowPrint(bottles.toString() + lines[0], SPEED)
        await sleep(LINE_PAUSE);
        await slowPrint(bottles.toString() + lines[1], SPEED)
        await sleep(LINE_PAUSE);
        await slowPrint(lines[2], SPEED)
        await sleep(LINE_PAUSE);
        bottles = bottles - 1 ; // Decrease the number of bottles by one.

        if (bottles > 0) {  // Print the last line of the current stanza.
            await slowPrint(bottles.toString() + lines[3], SPEED)
        } else {  // Print the last line of the entire song.
            await slowPrint('No more bottles of milk on the wall!', SPEED)
        }

        await sleep(LINE_PAUSE);
        console.log();  // Print a newline.

        // Choose a random line to make "sillier":
        let lineNum = Math.floor(Math.random() * 4);

        let line = lines[lineNum];
        let effect = Math.floor(Math.random() * 4);

        if (effect === 0) {  // Replace a character with a space.
            let charIndex = Math.floor(Math.random() * line.length);
            line = replaceAt(line, charIndex, ' ');
        } else if (effect === 1) {  // Change the casing of a character.
            let charIndex = Math.floor(Math.random() * line.length);
            if (line[charIndex].toUpperCase() == line[charIndex]) {
                line = replaceAt(line, charIndex, line[charIndex].toLowerCase());
            } else if (line[charIndex].toLowerCase() == line[charIndex]) {
                line = replaceAt(line, charIndex, line[charIndex].toUpperCase());
            }
        } else if (effect === 2) {  // Transpose two characters.
            let charIndex = Math.floor(Math.random() * (line.length - 1));
            let firstChar = line[charIndex]
            let secondChar = line[charIndex + 1]
            line = replaceAt(line, charIndex, secondChar);
            line = replaceAt(line, charIndex + 1, firstChar);
        } else if (effect === 3) {  // Double a character.
            let charIndex = Math.floor(Math.random() * (line.length - 1));
            line = line.substr(0, charIndex) + line[charIndex] + line.substr(charIndex);
        }

        lines[lineNum] = line;  // Update the strings in lines.
    }
}

main();

File: src/gamesbyexample/ninetyninebottles2.py
"""niNety-nniinE BoOttels of Mlik On teh waLl
By Al Sweigart al@inventwithpython.com
Print the full lyrics to one of the longest songs ever! The song
gets sillier and sillier with each verse. Press Ctrl-C to stop.
This code is available at https://nostarch.com/big-book-small-python-programming
Tags: short, scrolling, word"""

import random, sys, time

# Set up the constants:
# (!) Try changing both of these to 0 to print all the lyrics at once.
SPEED = 0.01  # The pause in between printing letters.
LINE_PAUSE = 1.5  # The pause at the end of each line.


def slowPrint(text, pauseAmount=0.1):
    """Slowly print out the characters in text one at a time."""
    for character in text:
        # Set flush=True here so the text is immediately printed:
        print(character, flush=True, end='')  # end='' means no newline.
        time.sleep(pauseAmount)  # Pause in between each character.
    print()  # Print a newline.


print('niNety-nniinE BoOttels, by Al Sweigart al@inventwithpython.com')
print()
print('(Press Ctrl-C to quit.)')

time.sleep(2)

bottles = 99  # This is the starting number of bottles.

# This list holds the string used for the lyrics:
lines = [' bottles of milk on the wall,',
         ' bottles of milk,',
         'Take one down, pass it around,',
         ' bottles of milk on the wall!']

try:
    while bottles > 0:  # Keep looping and display the lyrics.
        slowPrint(str(bottles) + lines[0], SPEED)
        time.sleep(LINE_PAUSE)
        slowPrint(str(bottles) + lines[1], SPEED)
        time.sleep(LINE_PAUSE)
        slowPrint(lines[2], SPEED)
        time.sleep(LINE_PAUSE)
        bottles = bottles - 1  # Decrease the number of bottles by one.

        if bottles > 0:  # Print the last line of the current stanza.
            slowPrint(str(bottles) + lines[3], SPEED)
        else:  # Print the last line of the entire song.
            slowPrint('No more bottles of milk on the wall!', SPEED)

        time.sleep(LINE_PAUSE)
        print()  # Print a newline.

        # Choose a random line to make "sillier":
        lineNum = random.randint(0, 3)

        # Make a list from the line string so we can edit it. (Strings
        # in Python are immutable.)
        line = list(lines[lineNum])

        effect = random.randint(0, 3)
        if effect == 0:  # Replace a character with a space.
            charIndex = random.randint(0, len(line) - 1)
            line[charIndex] = ' '
        elif effect == 1:  # Change the casing of a character.
            charIndex = random.randint(0, len(line) - 1)
            if line[charIndex].isupper():
                line[charIndex] = line[charIndex].lower()
            elif line[charIndex].islower():
                line[charIndex] = line[charIndex].upper()
        elif effect == 2:  # Transpose two characters.
            charIndex = random.randint(0, len(line) - 2)
            firstChar = line[charIndex]
            secondChar = line[charIndex + 1]
            line[charIndex] = secondChar
            line[charIndex + 1] = firstChar
        elif effect == 3:  # Double a character.
            charIndex = random.randint(0, len(line) - 2)
            line.insert(charIndex, line[charIndex])

        # Convert the line list back to a string and put it in lines:
        lines[lineNum] = ''.join(line)
except KeyboardInterrupt:
    sys.exit()  # When Ctrl-C is pressed, end the program.
